entity/c/c_api.c

```diff
@@ -0,0 +1,197 @@
+    unsigned char entity_nonce[HS_NONCE_SIZE];
+    unsigned int parsed_buf_length;
+    unsigned char *parsed_buf = parse_handshake_1(s_key, entity_nonce, &parsed_buf_length);
+    unsigned char sender[128]; // TODO: actually only needs 19 bytes.
+    unsigned int sender_length;
+    make_sender_buf(parsed_buf, parsed_buf_length, SKEY_HANDSHAKE_1, sender, &sender_length);
+    write(sock, sender, sender_length);
+    free(parsed_buf);
+    entity_client_state = HANDSHAKE_1_SENT;
+
+    // received handshake 2
+    unsigned char received_buf[1000];
+    unsigned int received_buf_length = read(sock, received_buf, sizeof(received_buf));
+    unsigned char message_type;
+    unsigned int data_buf_length;
+    unsigned char *data_buf = parse_received_message(received_buf, received_buf_length, &message_type, &data_buf_length);
+    if (message_type == SKEY_HANDSHAKE_2)
+    {
+        if (entity_client_state != HANDSHAKE_1_SENT)
+        {
+            printf("Comm init failed: wrong sequence of handshake, disconnecting...\n");
+        }
+        unsigned int parsed_buf_length;
+        unsigned char *parsed_buf = check_handshake_2_send_handshake_3(data_buf, data_buf_length, entity_nonce, s_key, &parsed_buf_length);
+        unsigned char sender[256];
+        unsigned int sender_length;
+        make_sender_buf(parsed_buf, parsed_buf_length, SKEY_HANDSHAKE_3, sender, &sender_length);
+        write(sock, sender, sender_length);
+        free(parsed_buf);
+        printf("switching to IN_COMM\n");
+        entity_client_state = IN_COMM;
+    }
+    sent_seq_num = 0;
+    st_time = 0;
+    printf("wait\n");
+    return sock;
+}
+
+session_key_t *server_secure_comm_setup(config_t *config, int clnt_sock)
+{
+    entity_server_state = IDLE;
+    unsigned char server_nonce[HS_NONCE_SIZE];
+    session_key_t *s_key;
+    while (1)
+    {
+        unsigned char received_buf[1024];
+        int received_buf_length = read(clnt_sock, received_buf, sizeof(received_buf));
+        unsigned char message_type;
+        unsigned int data_buf_length;
+        unsigned char *data_buf = parse_received_message(received_buf, received_buf_length, &message_type, &data_buf_length);
+        if (message_type == SKEY_HANDSHAKE_1)
+        {
+            printf("received session key handshake1\n");
+            if (entity_server_state != IDLE)
+            {
+                error_handling("Error during comm init - in wrong state, expected: IDLE, disconnecting...\n");
+            }
+            printf("switching to HANDSHAKE_1_RECEIVED state.\n");
+            entity_server_state = HANDSHAKE_1_RECEIVED;
+            unsigned char expected_key_id[SESSION_KEY_ID_SIZE];
+            memcpy(expected_key_id, data_buf, SESSION_KEY_ID_SIZE);
+            unsigned int expected_key_id_int = read_unsigned_int_BE(expected_key_id, SESSION_KEY_ID_SIZE);
+            /*
+            //TODO: Implement? need to think how.
```

Comment:
What is this TODO for?

Reply:
**THIS IS RELATED TO THE COMMENT ABOVE OF SENDER_BUF SIZE

Currently the send message does not support dynamic sizes, the max length is shorter than 1024 because this 1024 includes header, encryptions, hmacs etc.

When dynamic buffer sizes are needed, most of the code is implemented in malloc and free, but not when making the sender buf. Do you recommend to change the total connection part to dynamic memory allocation or only when sending messages?

In more details, for example, the auth_hello_reply message's size is dynamic due to the sender_buf(e.g. net1.client) and purpose_buf(e.g. {"group":"Servers"}, {"keyId": 00000000}). 
On the other hand, the buffer size when sending handshake1 from client to server is static(Actually changes by encryption scheme, but not implemented yet.). 

So, what do you recommend to use when sending bufs? Static of dynamic? If static how should I decide the size of the buffer.

Reply:
Need to check if the entity_server currently holds the session key of the expected_key_id.
If the entity_server already has the corresponding session key, it does not have to request session key from Auth.

Added comments to the TODO.

Reply:
Finished implementing in commit 6f9fd165859b9fc198f30757c6e0341de59fe17a .